File: 2017/go/day2/day02.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"

	"aoc/2017/arrays"
)

func Checksum(in string, f func([]int) int) int {
	rows := strings.Split(in, "\n")
	sum := 0
	for _, row := range rows {
		nums := convertRow(row)
		sum += f(nums)
	}
	return sum
}

func findDiff(nums []int) int {
	return arrays.Max(nums) - arrays.Min(nums)
}

func quotient(nums []int) int {
	for i, num := range nums {
		for j, div := range nums {
			if i == j {
				continue
			}
			if num % div == 0 {
				return num / div
			}
		}
	}
	return 0
}

func convertRow(row string) []int {
	nums := strings.Split(row, "\t")
	out := make([]int, len(nums))
	for i, num := range nums {
		n, err := strconv.Atoi(num)
		if err == nil {
			out[i] = n
		}
	}

	return out
}

func main() {
	input, err := ioutil.ReadFile("day2.in")
	if err == nil {
		in := string(input)
		fmt.Printf("Day 1 Part 1: %v\n", Checksum(in, findDiff))
		fmt.Printf("Day 1 Part 1: %v\n", Checksum(in, quotient))
	}
}

File: 2017/day02.rb
def checksum rows
  rows.map do |row|
    yield row
  end.reduce(:+)
end

def process input
  input.map do |row|
    row.split("\t").map { |x| x.strip.to_i }
  end
end

max_diff = Proc.new do |row|
  row.max - row.min
end

find_pair = Proc.new do |row|
  res = row.combination(2).select do |combo|
    combo.max % combo.min == 0
  end.first

  res.max / res.min
end

if $0 == __FILE__
  input = File.readlines('day2.in')
  rows = process input
  p "Part 1: #{checksum(rows, &max_diff)}"
  p "Part 1: #{checksum(rows, &find_pair)}"
end
